plugins/inputs/win_perf_counters/win_perf_counters.go

```diff
@@ -183,72 +250,22 @@ func (m *Win_PerfCounters) Gather(acc telegraf.Accumulator) error {
 		}
 	}
 
-	var bufSize uint32
-	var bufCount uint32
-	var size uint32 = uint32(unsafe.Sizeof(PDH_FMT_COUNTERVALUE_ITEM_DOUBLE{}))
-	var emptyBuf [1]PDH_FMT_COUNTERVALUE_ITEM_DOUBLE // need at least 1 addressable null ptr.
+	var emptyBuf [1]uint32 // need at least 1 addressable null ptr.
+	var counterValue PDH_FMT_COUNTERVALUE_DOUBLE
 
 	// For iterate over the known metrics and get the samples.
 	for _, metric := range m.itemCache {
 		// collect
 		ret := PdhCollectQueryData(metric.handle)
 		if ret == ERROR_SUCCESS {
-			ret = PdhGetFormattedCounterArrayDouble(metric.counterHandle, &bufSize,
-				&bufCount, &emptyBuf[0]) // uses null ptr here according to MSDN.
-			if ret == PDH_MORE_DATA {
-				filledBuf := make([]PDH_FMT_COUNTERVALUE_ITEM_DOUBLE, bufCount*size)
-				if len(filledBuf) == 0 {
-					continue
-				}
-				ret = PdhGetFormattedCounterArrayDouble(metric.counterHandle,
-					&bufSize, &bufCount, &filledBuf[0])
-				for i := 0; i < int(bufCount); i++ {
-					c := filledBuf[i]
-					var s string = UTF16PtrToString(c.SzName)
-
-					var add bool
-
-					if metric.include_total {
-						// If IncludeTotal is set, include all.
-						add = true
-					} else if metric.instance == "*" && !strings.Contains(s, "_Total") {
-						// Catch if set to * and that it is not a '*_Total*' instance.
-						add = true
-					} else if metric.instance == s {
-						// Catch if we set it to total or some form of it
-						add = true
-					} else if strings.Contains(metric.instance, "#") && strings.HasPrefix(metric.instance, s) {
-						// If you are using a multiple instance identifier such as "w3wp#1"
-						// phd.dll returns only the first 2 characters of the identifier.
-						add = true
-						s = metric.instance
-					} else if metric.instance == "------" {
-						add = true
-					}
-
-					if add {
-						fields := make(map[string]interface{})
-						tags := make(map[string]string)
-						if s != "" {
-							tags["instance"] = s
-						}
-						tags["objectname"] = metric.objectName
-						fields[sanitizedChars.Replace(metric.counter)] =
-							float32(c.FmtValue.DoubleValue)
-
-						measurement := sanitizedChars.Replace(metric.measurement)
-						if measurement == "" {
-							measurement = "win_perf_counters"
-						}
-						acc.AddFields(measurement, fields, tags)
-					}
-				}
-
-				filledBuf = nil
-				// Need to at least set bufSize to zero, because if not, the function will not
-				// return PDH_MORE_DATA and will not set the bufSize.
-				bufCount = 0
-				bufSize = 0
+			ret = PdhGetFormattedCounterValueDouble(metric.counterHandle, &emptyBuf[0], &counterValue)
```

Comment:
Sure, the previous code used PdhGetFormattedCounterArrayDouble, my version expands the counter queries to a list of individual counter queries and handle all the removed logic in different parts now. I also sanitize the names in AddItem().

```	
        sanitized_measurement := sanitizedChars.Replace(measurement)
	if sanitized_measurement == "" {
		sanitized_measurement = "win_perf_counters"
	}

	sanitized_counter := sanitizedChars.Replace(counter)
```